forwarder/writer_test.go

```diff
@@ -0,0 +1,874 @@
+		},
+		{
+			desc: "all expected labels present in end of labels slice",
+			arg: &remote.TimeSeries{
+				Labels: []*remote.LabelPair{
+					&remote.LabelPair{Name: model.MetricsPathLabel, Value: "/metrics"},
+					&remote.LabelPair{Name: model.MetricNameLabel, Value: "fake"},
+					&remote.LabelPair{Name: model.AlertNameLabel, Value: "boo"},
+					&remote.LabelPair{Name: model.JobLabel, Value: "foobar"},
+					&remote.LabelPair{Name: model.AddressLabel, Value: "sammy.example.com:9999"},
+					&remote.LabelPair{Name: model.InstanceLabel, Value: "inst8888"},
+				},
+			},
+			expected: "foobar-inst8888",
+		},
+		{
+			desc: "only job and address labels present in end of labels slice",
+			arg: &remote.TimeSeries{
+				Labels: []*remote.LabelPair{
+					&remote.LabelPair{Name: model.MetricsPathLabel, Value: "/metrics"},
+					&remote.LabelPair{Name: model.MetricNameLabel, Value: "fake"},
+					&remote.LabelPair{Name: model.AlertNameLabel, Value: "boo"},
+					&remote.LabelPair{Name: model.JobLabel, Value: "foobar"},
+					&remote.LabelPair{Name: model.AddressLabel, Value: "sammy.example.com:9999"},
+				},
+			},
+			expected: "foobar-sammy.example.com:9999",
+		},
+		{
+			desc: "only job and address labels present in beginning of labels slice",
+			arg: &remote.TimeSeries{
+				Labels: []*remote.LabelPair{
+					&remote.LabelPair{Name: model.JobLabel, Value: "foobar"},
+					&remote.LabelPair{Name: model.AddressLabel, Value: "sammy.example.com:9999"},
+					&remote.LabelPair{Name: model.MetricsPathLabel, Value: "/metrics"},
+					&remote.LabelPair{Name: model.MetricNameLabel, Value: "fake"},
+					&remote.LabelPair{Name: model.AlertNameLabel, Value: "boo"},
+				},
+			},
+			expected: "foobar-sammy.example.com:9999",
+		},
+		{
+			desc: "all expected labels present, but instance label before address label",
+			arg: &remote.TimeSeries{
+				Labels: []*remote.LabelPair{
+					&remote.LabelPair{Name: model.JobLabel, Value: "foobar"},
+					&remote.LabelPair{Name: model.InstanceLabel, Value: "inst8888"},
+					&remote.LabelPair{Name: model.MetricsPathLabel, Value: "/metrics"},
+					&remote.LabelPair{Name: model.MetricNameLabel, Value: "fake"},
+					&remote.LabelPair{Name: model.AddressLabel, Value: "sammy.example.com:9999"},
+					&remote.LabelPair{Name: model.AlertNameLabel, Value: "boo"},
+				},
+			},
+			expected: "foobar-inst8888",
+		},
+		{
+			desc: "only job and instance labels",
+			arg: &remote.TimeSeries{
+				Labels: []*remote.LabelPair{
+					&remote.LabelPair{Name: model.JobLabel, Value: "foobar"},
+					&remote.LabelPair{Name: model.MetricsPathLabel, Value: "/metrics"},
+					&remote.LabelPair{Name: model.MetricNameLabel, Value: "fake"},
+					&remote.LabelPair{Name: model.InstanceLabel, Value: "inst8888"},
+					&remote.LabelPair{Name: model.AlertNameLabel, Value: "boo"},
+				},
+			},
+			expected: "foobar-inst8888",
+		},
+	}
+
+	for i, test := range happyPathTests {
+		t.Logf("happy path tests %d: %q", i, test.desc)
```

Comment:
Now that 1.7 is out, I'd recommend using subtests, since you can choose to run a single test table case with them if needed.